src/main/java/in/org/projecteka/hiu/dataflow/HealthInfoManager.java

```diff
@@ -35,6 +45,62 @@
                                 consentDetail.get("hipName"))));
     }
 
+    public Flux<PatientDataEntry> fetchHealthInformation(List<String> dataRequestIds, String requesterId,
+                                                         int limit, int offset) {
+        return patientConsentRepository.fetchConsentRequestIds(dataRequestIds)
+                .collectList()
+                .flatMapMany(dataFlowRepository::fetchDataPartDetails)
+                .collectList()
+                .filter(dataParts -> isValidRequester(dataParts, requesterId))
+                .switchIfEmpty(Mono.error(unauthorizedRequester()))
+                .flatMapMany(dataParts -> getDataEntries(limit, offset, dataParts));
+    }
+
+    private boolean isValidRequester(List<DataPartDetail> dataParts, String requesterId) {
+        return dataParts.stream().allMatch(dataPart -> dataPart.getRequester().equals(requesterId));
+    }
+
+    private Flux<PatientDataEntry> getDataEntries(int limit, int offset, List<DataPartDetail> dataParts) {
+        HashMap<String, List<HealthInfoStatus>> dataPartStatuses = new HashMap<>();
+        HashMap<String, PatientDataEntry.PatientDataEntryBuilder> dataEntries = new HashMap<>();
+        dataParts.forEach(dataPart -> {
+            var statuses = dataPartStatuses.get(dataPart.getTransactionId());
+            dataEntries.put(dataPart.getTransactionId(), PatientDataEntry.builder()
+                    .consentRequestId(dataPart.getConsentRequestId())
+                    .hipId(dataPart.getHipId())
+                    .consentArtefactId(dataPart.getConsentArtifactId()));
+            if (statuses == null) {
+                dataPartStatuses.put(dataPart.getTransactionId(), List.of(dataPart.getStatus()));
+                return;
+            }
+            statuses.add(dataPart.getStatus());
+            dataPartStatuses.put(dataPart.getTransactionId(), statuses);
+        });
+        List<String> transactionIds = List.copyOf(dataEntries.keySet());
+        List<PatientDataEntry> partsInProcess = getDataPartsInProcess(dataPartStatuses, dataEntries);
+        return healthInformationRepository.getHealthInformation(transactionIds, limit, offset)
+                .map(healthInfo -> dataEntries.get(healthInfo.get("transaction_id").toString())
+                        .status(toStatus((String) healthInfo.get("status")))
+                        .data(healthInfo.get("data")).build())
+                .mergeWith(Flux.fromIterable(partsInProcess));
+    }
+
+    private List<PatientDataEntry> getDataPartsInProcess(HashMap<String, List<HealthInfoStatus>> dataPartStatuses,
+                                                         HashMap<String, PatientDataEntry.PatientDataEntryBuilder> dataEntries) {
```

Comment:
I don't think you need to pass this additional map HashMap\<String, PatientDataEntry.PatientDataEntryBuilder> dataEntries. Think about it.

Reply:
done. Please check.